typeText replace option doesn't work when the input already has a value that fit maxlength

### What is your Test Scenario?
Replacing the old value in the text input with a new value. The old value has a length that fit maxlength.
 
### What is the Current behavior?
The new value doesn't replace the old value.
 
### What is the Expected behavior?
The new value replace the old value.
 
### What is your web application and your TestCafe test code?
 
<details>
<summary>Your website URL (or attach your complete example):</summary>
The example app is just a simple html form.

```html
<form>
    <input type="text" id="name" maxlength="5">
</form>
```

</details>

<details>
<summary>Your complete test code (or attach your test files):</summary>

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `file:///path/to/form.html`;

test('Example test', async t => {
    await t
        .typeText('#name', '12345') // Input the old value.
        .typeText('#name', 'abcde', { replace: true }) // Replace the new value.
        .expect(Selector('#name').value).eql('abcde'); // Assert the input value.
});
```
</details>

<details>
<summary>Your complete configuration file (if any):</summary>

Not using configuration file.
</details>

<details>
<summary>Your complete test report:</summary>
Actually it's not getting any error.

```
 Running tests in:
 - Chrome 88.0.4324.146 / macOS 10.15.7

 Example
 ✖ Example test

   1) AssertionError: expected '12345' to deeply equal 'abcde'
      
      + expected - actual
      
      -'12345'
      +'abcde'
```
</details>

<details>
<summary>Screenshots:</summary>
<!-- If applicable, add screenshots to help explain the issue. -->

```

```
</details>

### Steps to Reproduce:

1. Create an example html file.
```html:form.html
<form>
    <input type="text" id="name" maxlength="5">
</form>
```
2. Create an example test file.
```js:example.js
import { Selector } from 'testcafe';

fixture `Example`
    .page `file:///path/to/form.html`;

test('Example test', async t => {
    await t
        .typeText('#name', '12345') // Input the old value.
        .typeText('#name', 'abcde', { replace: true }) // Replace the new value.
        .expect(Selector('#name').value).eql('abcde'); // Assert the input value.
});
```
3. Execute this command
```
testcafe chrome example.js
```
4. Check the behavior.
 
### Your Environment details:
 
* testcafe version: 1.10.1 
* node.js version: v14.6.0  
* command-line arguments: testcafe chrome example.js
* browser name and version: Chrome 88.0.4324.146 
* platform and version: macOS 10.15.7

--- a/src/client/automation/playback/type/type-text.js
+++ b/src/client/automation/playback/type/type-text.js
@@ -263,15 +263,16 @@ function _typeTextToTextEditable (element, text) {
     if (elementMaxLength < 0)
         elementMaxLength = browserUtils.isIE && browserUtils.version < 17 ? 0 : null;
 
-    if (elementMaxLength === null || isNaN(elementMaxLength) || elementMaxLength > elementValue.length) {
+    const newElementValue = elementValue.substring(0, startSelection) + text + elementValue.substring(endSelection, elementValue.length);
+
+    if (elementMaxLength === null || isNaN(elementMaxLength) || elementMaxLength >= newElementValue.length) {
         // NOTE: B254013
         if (isInputTypeNumber && browserUtils.isIOS && elementValue[elementValue.length - 1] === '.') {
             startSelection += 1;
             endSelection += 1;
         }
 
-        domUtils.setElementValue(element, elementValue.substring(0, startSelection) + text +
-                          elementValue.substring(endSelection, elementValue.length));
+        domUtils.setElementValue(element, newElementValue);
 
         textSelection.select(element, startSelection + textLength, startSelection + textLength);
     }

--- a/test/functional/fixtures/regression/gh-5921/pages/index.html
+++ b/test/functional/fixtures/regression/gh-5921/pages/index.html
@@ -1,8 +1,8 @@
 <!DOCTYPE html>
 <html>
 <head>
-    <input type="text" id="input" maxlength="4" value="test">
 </head>
 <body>
+    <input type="text" id="input" maxlength="4" value="test">
 </body>
 </html>

--- a/test/functional/fixtures/regression/gh-5921/pages/index.html
+++ b/test/functional/fixtures/regression/gh-5921/pages/index.html
@@ -1,0 +1,8 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <input type="text" id="input" maxlength="4" value="test">
+</head>
+<body>
+</body>
+</html>

--- a/test/functional/fixtures/regression/gh-5921/test.js
+++ b/test/functional/fixtures/regression/gh-5921/test.js
@@ -1,0 +1,5 @@
+describe('[Regression](GH-5921) typeText should replace the old value when the old value length fits maxlength', () => {
+    it('Should replace the old value when it fits the maxlength', () => {
+        return runTests('testcafe-fixtures/index.js');
+    });
+});

--- a/test/functional/fixtures/regression/gh-5921/testcafe-fixtures/index.js
+++ b/test/functional/fixtures/regression/gh-5921/testcafe-fixtures/index.js
@@ -1,0 +1,10 @@
+import { Selector } from 'testcafe';
+
+fixture `GH-5921 - typeText should replace the old value when the old value length fits maxlength`
+    .page `http://localhost:3000/fixtures/regression/gh-5921/pages/index.html`;
+
+test(`Replace the old value`, async t => {
+    await t
+        .typeText('#input', 'cafe', { replace: true })
+        .expect(Selector('#input').value).eql('cafe');
+});

